zero-code-instrumentation/lib/zero-code-instrumentation.rb

```diff
@@ -0,0 +1,140 @@
+    'ethon' => 'OpenTelemetry::Instrumentation::Ethon',
+    'excon' => 'OpenTelemetry::Instrumentation::Excon',
+    'faraday' => 'OpenTelemetry::Instrumentation::Faraday',
+    'grape' => 'OpenTelemetry::Instrumentation::Grape',
+    'graphql' => 'OpenTelemetry::Instrumentation::GraphQL',
+    'http_client' => 'OpenTelemetry::Instrumentation::HttpClient',
+    'mongo' => 'OpenTelemetry::Instrumentation::Mongo',
+    'mysql2' => 'OpenTelemetry::Instrumentation::Mysql2',
+    'net_http' => 'OpenTelemetry::Instrumentation::Net::HTTP',
+    'pg' => 'OpenTelemetry::Instrumentation::PG',
+    'que' => 'OpenTelemetry::Instrumentation::Que',
+    'racecar' => 'OpenTelemetry::Instrumentation::Racecar',
+    'rack' => 'OpenTelemetry::Instrumentation::Rack',
+    'rails' => 'OpenTelemetry::Instrumentation::Rails',
+    'rake' => 'OpenTelemetry::Instrumentation::Rake',
+    'rdkafka' => 'OpenTelemetry::Instrumentation::Rdkafka',
+    'redis' => 'OpenTelemetry::Instrumentation::Redis',
+    'restclient' => 'OpenTelemetry::Instrumentation::RestClient',
+    'resque' => 'OpenTelemetry::Instrumentation::Resque',
+    'ruby_kafka' => 'OpenTelemetry::Instrumentation::RubyKafka',
+    'sidekiq' => 'OpenTelemetry::Instrumentation::Sidekiq',
+    'sinatra' => 'OpenTelemetry::Instrumentation::Sinatra'
+  }.freeze
+
+  def detect_resource_from_env
+    env = ENV['OTEL_RUBY_RESOURCE_DETECTORS'].to_s
+    additional_resource = ::OpenTelemetry::SDK::Resources::Resource.create({})
+
+    env.split(',').each do |detector|
+      case detector
+      when 'container'
+        additional_resource = additional_resource.merge(::OpenTelemetry::Resource::Detector::Container.detect) if defined? ::OpenTelemetry::Resource::Detector::Container
+      when 'google_cloud_platform'
+        additional_resource = additional_resource.merge(::OpenTelemetry::Resource::Detector::GoogleCloudPlatform.detect) if defined? ::OpenTelemetry::Resource::Detector::GoogleCloudPlatform
+      when 'azure'
+        additional_resource = additional_resource.merge(::OpenTelemetry::Resource::Detector::Azure.detect) if defined? ::OpenTelemetry::Resource::Detector::Azure
+      end
+    end
+
+    additional_resource
+  end
+
+  def determine_enabled_instrumentation
+    env = ENV['OTEL_RUBY_ENABLED_INSTRUMENTATIONS'].to_s
+
+    env.split(',').map { |instrumentation| OTEL_INSTRUMENTATION_MAP[instrumentation] }
+  end
+
+  def require_otel
+    lib = File.expand_path('..', __dir__)
+    $LOAD_PATH.reject! { |path| path.include?('zero-code-instrumentation') }
+    $LOAD_PATH.unshift(lib)
+
+    begin
+      required_instrumentation = determine_enabled_instrumentation
+
+      OpenTelemetry::SDK.configure do |c|
+        c.resource = detect_resource_from_env
+        if required_instrumentation.empty?
+          c.use_all # enables all instrumentation!
+        else
+          required_instrumentation.each do |instrumentation|
+            c.use instrumentation
+          end
+        end
```

Comment:
I think what I would prefer is having this code run `use_all(instrumentation_config)` as opposed to installing each independenty. 

E.g. 

```ruby
  def determine_enabled_instrumentation
    env = ENV['OTEL_RUBY_ENABLED_INSTRUMENTATIONS'].to_s

    env.split(',').each_with_object({}) { |instrumentation, config| config[OTEL_INSTRUMENTATION_MAP[instrumentation]] = {} }
  end
```

This will create a structure that maps an instrumentation to a config hash and you can remove all of thie conditional logic. 

```ruby
  c.use_all(required_instrumentation)
```

Reply:
I have tested the `use_all(required_instrumentation)`. I think `use_all` will still try to install all required instrumentation (e.g. required from `require 'opentelemetry-instrumentation-all'`) and ignore OTEL_RUBY_ENABLED_INSTRUMENTATIONS. 

Registry `@instrumentation` ([L64](https://github.com/open-telemetry/opentelemetry-ruby/blob/main/registry/lib/opentelemetry/instrumentation/registry.rb#L64)) will determine which instrumentation will be installed.